mgr/dashboard: Add minimalistic browsable API

sebastian-philipp opened the pull request:
Also provides a simple HTML form to POST
data to a `RESTController`'s `create()` method.

obligatory Screenshot:

![html-debug-view](https://user-images.githubusercontent.com/2574405/37527362-29fbe794-2932-11e8-8804-66bd175cb033.png)



Signed-off-by: Sebastian Wagner <sebastian.wagner@suse.com>

rjfd commented:
This HTML form generator looks nice and helpful to test some API endpoints during development, but I have one question on how are we supposed to use this? should we add the `@debug_html_page` decorator while developing a rest controller endpoint, but then remove the decorator when we push the commits to the PR? or should we leave the decorator in the code? (if it's the latter then we need some config option to disable the debug page in production)

sebastian-philipp commented:
@rjfd actually, I would keep it enabled also for production use.

1. If you look at a page, if is obvious that this is not meant to be used in production
2. It doesn't provide anything that you cannot use via curl. 
3. In openATTIC, we also didn't disable this API viewer. 

Another idea would be to rephrase this as preliminary "API documentation" or "(incomplete) browsable API".

LenzGr commented:
I like the basic idea behind this, but if it's supposed to be enabled by default, the name `@debug_html_page` is indeed misleading.

Having an automated way of creating a browseable "API reference" is actually a nice side benefit that makes me want to have this :)

In Django, this is a builtin-feature in the Django REST framework (IIRC), where it can be enabled/disabled on a global level. Maybe this would be a better approach than having to add a decorator to every controller endpoint?

sebastian-philipp commented:
@rjfd : Just added `ENABLE_BROWSABLE_API` setting to our settings. 
@LenzGr: Renamed the PR to "browsable API"

I'd favor this setting to be `True` by default, but I don't mind to set this to False

sebastian-philipp commented:
@LenzGr: added a [disclaimer](https://github.com/ceph/ceph/pull/20873/files#diff-baf1e07da5b5ebf5293ee77bf0364f04R113)

sebastian-philipp commented:
jenkins retest this please

sebastian-philipp commented on src/pybind/mgr/dashboard/tools.py:
```diff
@@ -473,27 +636,27 @@ def stop(cls):
+    def register(cls, func, n_types=None):
         """Registers function to listen for notifications
 
         If the second parameter `types` is omitted, the function in `func`
```
fixed

rjfd commented on src/pybind/mgr/dashboard/tools.py:
```diff
@@ -473,27 +617,27 @@ def stop(cls):
 
     @classmethod
-    def register(cls, func, types=None):
+    def register(cls, func, n_types=None):
```
Can you put this renaming part into a separate commit?

sebastian-philipp replied:
fixed

rjfd commented on src/pybind/mgr/dashboard/controllers/monitor.py:
```diff
@@ -14,7 +14,7 @@
     @cherrypy.expose
     @cherrypy.tools.json_out()
-    def default(self):
+    def default(self, *_vpath, **_params):
```
Is there a way to not require to add those parameters in all `default` methods?

sebastian-philipp replied:
Pylint says; `W:464, 4: Parameters differ from overridden 'default' method (arguments-differ)`. In general, I think, it's OK to have consistent method arguments for derived classes.

rjfd commented:
QA tests run successful:
http://pulpito.ceph.com/rdias-2018-03-26_16:07:05-rados:mgr-wip-rdias-testing-distro-basic-smithi/
http://pulpito.ceph.com/rdias-2018-03-26_18:46:21-rados-wip-rdias-testing-distro-basic-smithi/
